apps/aecore/src/aec_fork_signalling.erl

```diff
@@ -0,0 +1,350 @@
+        undefined ->
+            send_server_msg(Pid, compute),
+            Results1 = add_result(BlockHash, LastSigBlockHash, Results),
+            Results2 = add_result(LastSigBlockHash, pending, Results1),
+            State#state{results = Results2}
+    end;
+handle_worker_msg(_Worker, {add_result, BlockHash, Result},
+                  #state{results = Results} = State) ->
+    State#state{results = add_result(BlockHash, Result, Results)}.
+
+handle_down(Worker, _Rsn, #state{workers = Workers} = State) ->
+    State#state{workers = del_worker(Worker, Workers)}.
+
+get_action_from_block_height(Block, Fork) ->
+    case {is_last_signalling_block(Block, Fork),
+          is_signalling_result_required(Block, Fork)} of
+        {true, _}      -> spawn_worker;
+        {false, true}  -> set_prev_result_or_spawn_worker;
+        {false, false} -> skip_block
+    end.
+
+spawn_worker(Block, BlockHash, #state{fork = Fork, workers = Workers} = State) ->
+    Worker = new_worker([Block, BlockHash, Fork]),
+    State#state{workers = add_worker(Worker, BlockHash, Workers)}.
+
+kill_worker({Pid, Ref} = Worker, #state{workers = Workers} = State) ->
+    cancel_monitor(Ref),
+    exit(Pid, shutdown),
+    flush_worker_msgs(),
+    State#state{workers = del_worker(Worker, Workers)}.
+
+set_prev_result_or_spawn_worker(Block, BlockHash, #state{results = Results} = State) ->
+    PrevKeyHash = aec_blocks:prev_key_hash(Block),
+    case get_result(PrevKeyHash, Results) of
+        R when ?IS_RESULT(R) ->
+            %% The PrevKeyHash key is associated with a result (PrevKeyhash is
+            %% a block hash of block at height HE - 1). The BlockHash is
+            %% set to point to the PrevKeyHash (so it's possible to retrieve
+            %% the result for BlockHash).
+            State#state{results = add_result(BlockHash, PrevKeyHash, Results)};
+        P when ?IS_POINTER(P) ->
+            %% The PrevKeyhash is associated with a pointer that is a block
+            %% hash at height HE - 1, which has the result. The BlockHash is
+            %% set to point to the same block hash at height HE - 1.
+            %% The PrevKeyhash associated with the pointer can now be discarded
+            %% to avoid caching results for all the blocks between HE and
+            %% H2 - 1 (inclusive). Only the block hash at height H2 is
+            %% important.
+            Results1 = add_result(BlockHash, P, Results),
+            Results2 = del_result(PrevKeyHash, Results1),
+            State#state{results = Results2};
+        undefined ->
+            %% There is no result reachable via PrevKeyhash. A new process is
+            %% spawned, it will set both a pointer (for BlockHash) and
+            %% a result for a block hash at height HE - 1.
+            spawn_worker(Block, BlockHash, State)
+    end.
+
+worker_process(Parent, Block, BlockHash, Fork) ->
+    case search_last_signalling_block(Block, BlockHash, Fork) of
```

Comment:
I see this used that performs several distinct small DB transactions. I expect:
- Not one large DB transaction for preventing locking unnecessarily.
- Not dirty as it would not play well with (async-)transaction.

Seems ok.